autoscheduler/frontend/src/components/Schedule/Schedule.tsx

```diff
@@ -1,25 +1,252 @@
+  const schedule = useSelector<RootState, Meeting[]>((state) => state.meetings);
+  const availabilityList = useSelector<RootState, Availability[]>((state) => state.availability);
+  const availabilityMode = useSelector<RootState, AvailabilityType>(
+    (state) => state.availabilityMode,
+  );
+  const selectedAvailability = useSelector<RootState, AvailabilityArgs>(
+    (state) => state.selectedAvailability,
+  );
+  const dispatch = useDispatch();
+
+  /* state */
+  const [startDay, setStartDay] = React.useState(null);
+  // either the start or end time, in minutes since midnight,
+  // for the availability currently being added
+  const [time1, setTime1] = React.useState(null);
 
-  // helper functions for formatting
+  // state management for time display
+  const [hoveredDay, setHoveredDay] = React.useState(null);
+  const [mouseY, setMouseY] = React.useState<number>(null);
+  const [hoveredTime, setHoveredTime] = React.useState<number>(null);
+
+  // helper functions
   function formatHours(hours: number): number {
     return ((hours - 1) % 12) + 1;
   }
 
+  /**
+   * Given a MouseEvent in a calendar day, calculates the time, in minutes since midnight
+   * and rounded to the nearest 10, at which the mouse event was emitted
+   * @param evt
+   */
+  function eventToTime(evt: React.MouseEvent<HTMLDivElement, MouseEvent>): number {
+    const totalY = evt.currentTarget.clientHeight;
+    const yPercent = (evt.clientY - evt.currentTarget.getBoundingClientRect().top) / totalY;
+    const minutesPerDay = (LAST_HOUR - FIRST_HOUR) * 60;
+    const yMinutes = yPercent * minutesPerDay;
+    const roundedMinutes = Math.round(yMinutes / 10) * 10;
+    return roundedMinutes + FIRST_HOUR * 60;
+  }
+
+  /**
+   * Using time1 and time2 from the given availability, returns arguments for
+   * an availability that is at least 30 minutes long and ends before 9 PM
+   */
+  function roundUpAvailability(avl: AvailabilityArgs): AvailabilityArgs[] {
+    const blockSize = Math.abs(avl.time2 - avl.time1);
+    if (blockSize < 30) {
+      if (avl.time2 < 20 * 60 + 30) {
+        if (avl.time1 > 8 * 60 + 30) {
+          // if the availability is solidly between 8:30 and 20:30, just round up
+          return [{
+            ...avl,
+            // if the sign is zero, then assumes positive by default
+            time2: avl.time1 + 30 * (Math.sign(avl.time2 - avl.time1) || 1),
+          }];
+        }
+
+        // if availability is close to the eddges, force it to 8 to 8:30 in 2 steps
+        return [{
+          ...avl,
+          time2: 8 * 60,
+        }, {
+          ...avl,
+          time1: 8 * 60,
+          time2: 8 * 60 + 30,
+        }];
+      }
+      // new time blocks cannot be later than 9 PM / 2100
+      return [{
+        ...avl,
+        time2: 21 * 60,
+      }, {
+        ...avl,
+        time1: 21 * 60,
+        time2: 20 * 60 + 30,
+      }];
+    }
+
+    // if there are no problems, just use avl as is
+    return [avl];
+  }
```

Comment:
This looks more or less okay to me, but I don't want to accidentally break this, so I'm going to move this into `Availability.ts`, add unit tests, and then apply the suggested changes. Just letting you know so you know where to look

Reply:
Oh wait, on second thought, I don't think this is going to work due to the assumption that `time2` is always after `time1`. When updating availabilities, `time1` actually needs to be whichever time is not being dragged, regardless of whether that is the start time or end time. Using the suggested code would result in `time1` sometimes being changed to the unstable, currently-dragged time, if the user was dragging toward the top, and this would cause a failed update, as I explained in a different comment.

This behavior is documented in the docstring for `updateAvailability`, but is incorrectly documented in the docstring for `AvailabilityArgs`. I have updated the documentation to reflect the correct usage.

Reply:
Okay, if you're not making this change then I think we should change the functionality slightly for the function you have: right now, if a block less than 30 minutes is created before 8:30, it will "snap" to a block from 8 to 8:30 which is inconsistent with any other time (keep start time and expand downwards). If you just remove the check for if it's before 8:30 then it behaves as I would expect. Is there a reason you decided to make it a special case?

Reply:
Yeah this was a bug that I reported I believe. If you start before 8:30 AM and drag upwards it automatically defaults to 30 minutes like we've had for awhile, so it'll force to 30 minutes before the start time (i.e. if you start at 8:20 AM and drag up, it'll default to 7:50 AM). What were you expecting for starting before 8:30 AM? Would you rather them default to 10-20 min / default to the start time being 8:00 AM regardless of what time we started it at?

Reply:
Ok nvm scratch that ^. Ryan's talking about starting from before 8:30 and dragging down (so it's the end time), while I'm talking about starting before 8:30 and dragging up (so it's the end time), but the two issues are related.

Likewise right now if you start at 8:10 and drag down, then delete it, it makes another availability from 8:00 AM to 8:30 AM.

Reply:
So let me clarify the expected behavior:

1. If you start at 8:20 and drag up to 8:10, it will expand to 8:00 - 8:30.
2. If you start at 8:10 and drag down to 8:20, it will expand to 8:10 - 8:40.

Would y'all agree?

Reply:
Yeah, since that's what happens with other times on the calendar I think that would be the best solution. Looking at it again, it seems like the same thing happens with times close to 9 PM (expanding to 8:30-9 rather than basing it on which way you drag), so that should be made consistent as well.